pkg/datasync/directory.go

```diff
@@ -0,0 +1,297 @@
+	bolt "go.etcd.io/bbolt"
+
+	cuckoo "github.com/panmari/cuckoofilter"
+	"github.com/samber/lo"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+func (s *Sync) syncDirectory(ctx context.Context, conn *grpc.ClientConn) error {
+	runStartTime := time.Now().UTC()
+	s.logger.Info().Str(status, started).Str("mode", s.options.Mode.String()).Msg(syncRun)
+
+	defer func() {
+		close(s.errChan)
+	}()
+
+	// error spew.
+	go func() {
+		for e := range s.errChan {
+			s.logger.Error().Err(e).Msg(syncRun)
+		}
+	}()
+
+	g := new(errgroup.Group)
+
+	g.Go(func() error {
+		err := s.subscriber(ctx)
+		if err != nil {
+			s.logger.Error().Err(err).Str(stage, "subscriber").Msg(syncRun)
+		}
+		return err
+	})
+
+	g.Go(func() error {
+		err := s.producer(ctx, conn)
+		if err != nil {
+			s.logger.Error().Err(err).Str(stage, "producer").Msg(syncRun)
+		}
+		return err
+	})
+
+	if err := g.Wait(); err != nil {
+		s.logger.Debug().Err(err).Msg(syncRun)
+	}
+
+	if Has(s.options.Mode, Diff) {
+		if err := s.diff(ctx); err != nil {
+			s.logger.Error().Err(err).Str(stage, "diff").Msg(syncRun)
+			return err
+		}
+	}
+
+	ts := <-s.tsChan
+	if err := s.setWatermark(ts); err != nil {
+		return err
+	}
+
+	runEndTime := time.Now().UTC()
+	s.logger.Info().Str(status, finished).Str("duration", runEndTime.Sub(runStartTime).String()).Msg(syncRun)
+	return nil
+}
+
+func (s *Sync) producer(ctx context.Context, conn *grpc.ClientConn) error {
+	s.logger.Info().Str(status, started).Msg(syncProducer)
+
+	var recvCtr, objCtr, relCtr int32
+
+	defer func() {
+		s.logger.Debug().Msg("producer closed export channel")
+		close(s.exportChan)
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	ts := &timestamppb.Timestamp{}
+
+	wm := s.getWatermark()
+	if Has(s.options.Mode, Watermark) {
+		ts = wm.Timestamp
+	}
+
+	if Has(s.options.Mode, Diff) {
+		s.filter = cuckoo.NewFilter(wm.getFilterSize())
+	}
+
+	s.logger.Debug().Str("start_from", ts.String()).Msg(syncProducer)
+	stream, err := dse3.NewExporterClient(conn).Export(ctx, &dse3.ExportRequest{
+		Options:   uint32(dse3.Option_OPTION_DATA),
+		StartFrom: ts,
+	})
+	if err != nil {
+		return err
+	}
+
+	for {
+		msg, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		atomic.AddInt32(&recvCtr, 1)
+
+		switch m := msg.Msg.(type) {
+		case *dse3.ExportResponse_Object:
+			atomic.AddInt32(&objCtr, 1)
+			if Has(s.options.Mode, Diff) {
+				s.filter.Insert(getObjectKey(m.Object))
+			}
+		case *dse3.ExportResponse_Relation:
+			atomic.AddInt32(&relCtr, 1)
+			if Has(s.options.Mode, Diff) {
+				s.filter.Insert(getRelationKey(m.Relation))
+			}
+		default:
+			s.logger.Debug().Msg("producer unknown message type")
+			continue // do not send msg to exportChan when unknown.
+		}
+
+		s.exportChan <- msg
+	}
+
+	s.logger.Info().Str(status, finished).
+		Int32("received", recvCtr).
```

Comment:
This should be `atomic.LoadInt32(&recvCtr)` but per my earlier comment, it's best to use the `atomic.Int32` type.